doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+have a corresponding test in :file:`mne/tests/test_evoked.py`.
+
+
+All new functionality must be documented
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This includes thorough docstring descriptions for all public API changes, as
+well as how-to examples or longer tutorials for major contributions. Docstrings
+for private functions may be more sparse, but should not be omitted.
+
+
+New API elements should be added to the master reference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Classes, functions, methods, and attributes cannot be cross-referenced unless
+they are included in the :doc:`python_reference`
+(:file:`doc/python_reference.rst`).
+
+
+Describe your changes in the changelog
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Include in your changeset a brief description of the change in the
+:doc:`changelog <whats_new>` (:file:`doc/whats_new.rst`; this can be skipped
+for very minor changes like correcting typos in the documentation). Note that
+there are sections of the changelog for each release, and separate subsections
+for bugfixes, new features, and changes to the public API. It is usually best
+to wait to add a line to the changelog until your PR is finalized, to avoid
+merge conflicts (since the changelog is updated with almost every PR).
+
+
+Test locally before opening pull requests (PRs)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+MNE-Python uses `continuous integration`_ (CI) to ensure code quality and
+test across multiple installation targets. However, the CIs are often slower
+than testing locally, especially when other contributors also have open PRs
+(which is basically always the case). Therefore, do not rely on the CIs to
+catch bugs and style errors for you; run the tests locally instead before
+opening a new PR and before each time you push additional changes to an
+already-open PR. See the :ref:`testing <run-tests>` section (below) for
+examples of how to run tests locally, and make sure you have the testing
+dataset installed.
+
+
+Make tests fast and thorough
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Whenever possible, use the testing dataset rather than one of the sample
+datasets when writing tests; it includes small versions of most MNE-Python
+objects (e.g., :class:`~mne.io.Raw` objects with short durations and few
+channels). You can also check which lines are missed by the tests, then modify
+existing tests (or write new ones) to target the missed lines. Here's an
+example that reports which lines within ``mne.viz`` are missed when running
+``test_evoked.py`` and ``test_topo.py``:
 
 .. code-block:: console
 
-   $ pip install sphinx sphinx-gallery sphinx_bootstrap_theme sphinx_fontawesome memory_profiler pillow
+    $ pytest --cov=mne.viz --cov-report=term-missing mne/viz/tests/test_evoked.py mne/viz/tests/test_topo.py
 
-and then within the ``mne/doc`` directory do:
+You can also use ``pytest --durations=20`` to ensure new or modified tests will
+not slow down the test suite too much.
+
```

Comment:
A critical missing piece, which should probably be toward the top of this list, is something like "Do not break existing code without good reason, and if you break it, do so properly", maybe succincly summarized as "Be careful and conservative about breaking user code". Things within this category:

- User API stability is an important consideration
- Things like renaming arguments that are slight misnomers are not worth the breakage
- Bug fixes (something not doing what it says it will do) do not require deprecation
- Changing the API (e.g., changing the default behavior of some function) requires deprecation so users have some warning where they can adapt their code